File: go/0454M-四数相加II.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-16 16:02:51
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	ans := 0
	m := make(map[int]int)
	for _, a := range nums1 {
		for _, b := range nums2 {
			m[a+b]++
		}
	}
	for _, a := range nums3 {
		for _, b := range nums4 {
			ans += m[-a-b]
		}
	}
	return ans
}

File: cpp/0454M-四数相加II.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-14 09:39:50
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3,
                     vector<int>& nums4) {
        unordered_map<int, int> set;
        for (int num1 : nums1) {
            for (int num2 : nums2) {
                set[num1 + num2]++;
            }
        }
        int ans = 0;
        for (int num3 : nums3) {
            for (int num4 : nums4) {
                if (set.find(-num3 - num4) != set.end()) {
                    ans += set[-num3 - num4];
                }
            }
        }
        return ans;
    }
};